org.eclipse.lsp4e/src/org/eclipse/lsp4e/format/IFormatRegions.java

```diff
@@ -0,0 +1,26 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   See git history
+ *******************************************************************************/
+package org.eclipse.lsp4e.format;
+
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.IRegion;
+
+public interface IFormatRegions {
+
+	/**
+	 * Get the formatting regions
+	 * @param document
+	 * @return region to be formatted or null if document should not be formatted on save.
```

Comment:
This gives the impression that this particular provider can cause the document to not be formatted at all by returning `null` even if some other suitable IFormatRegions for this document can contribute ranges to format.

Reply:
Because of the contributive nature of Eclipse Platform and OSGi service, 1 provider returning null doesn't mean that there is no provider that can apply. This is related to my former questions about the case of multiple extensions to handle different LS or documents: when this happens, we need something to pick the right one and use it; the right one may depend on various things.

Reply:
> 1 provider returning null doesn't mean that there is no provider that can apply.

There should be a single `IFormatRegionsProvider` provider for a LS definition. The filtering is done by the `serverDefinitionId` property in the OSGI component definition. This is my  `IFormatRegionsProvider` implementation for the CDT LS:
```java
@Component(property = { "serverDefinitionId:String=org.eclipse.cdt.lsp.server" })
public class FormatOnSave implements IFormatRegionsProvider {
	private final FormatEditedLines formatEditedLines = new FormatEditedLines();

	@Reference
	private EditorConfiguration configuration;

	@Override
	public IRegion[] getFormattingRegions(IDocument document) {
		var file = LSPEclipseUtils.getFile(document);
		if (file != null) {
			var editorOptions = configuration.options(file);
			if (editorOptions != null && editorOptions.formatOnSave()) {
				if (editorOptions.formatAllLines()) {
					return new IRegion[] { new Region(0, document.getLength()) };
				}
				if (editorOptions.formatEditedLines()) {
					return formatEditedLines.getFormattingRegions(document);
				}
			}
		}
		return null;
	}

}
```
>  This is related to my former questions about the case of multiple extensions to handle different LS or documents: when this happens, we need something to pick the right one and use it; the right one may depend on various things.

Please have a look at `org.eclipse.lsp4e.DocumentContentSynchronizer.getFormatRegions()` . There we pick up the `IFormatRegionsProvider` based on the `languageServerWrapper.serverDefinition.id` as filter:
`var serviceReferences = bundleContext.getAllServiceReferences(IFormatRegionsStrategyProvider.class.getName(), serverId);`

Reply:
Currently we pick up the first one, if trhere are several provider for the same LS id:
`reference = serviceReferences[0];`

>the right one may depend on various things.

If there is a use case for more than one `IFormatRegionsProvider` for a language server definition (e.g. for each LS  instance) then the filter could be modified.

Reply:
Maybe `IFormatRegionsProvider` could implement a method which returns the `serverID` for filtering purposes. IMO the filtering via OSGi  component property is more charming, since its a  one-liner: 
`var serviceReferences = bundleContext.getAllServiceReferences(IFormatRegionsProvider.class.getName(), serverId)`

I think it provides also more flexibility in the future. A new filter property can be easily added to the component definition without breaking the API.